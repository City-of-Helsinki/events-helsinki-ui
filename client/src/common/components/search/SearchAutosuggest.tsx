import React, { ChangeEvent, FunctionComponent } from "react";

import { ReactComponent as SearchIcon } from "../../../assets/icons/svg/search.svg";
import { Category as CategoryType } from "../../types";
import Category from "./Category";
import styles from "./searchAutosuggest.module.scss";

interface Props {
  categories: CategoryType[];
  onRemoveCategory: (category: CategoryType) => void;
  placeholder: string;
}

const SearchAutosuggest: FunctionComponent<Props> = ({
  categories,
  onRemoveCategory,
  placeholder
}) => {
  const container = React.useRef<HTMLDivElement | null>(null);
  const input = React.useRef<HTMLInputElement | null>(null);
  const [searchValue, setSearchValue] = React.useState("");

  const handleComponentClick = () => {
    if (input && input.current) {
      // Set focus on input field when clicking container
      input.current.focus();
    }
  };

  const handleInputChange = (event: ChangeEvent<HTMLInputElement>) => {
    setSearchValue(event.target.value);
  };

  React.useEffect(() => {
    const current = container && container.current;

    // Add event listener to component to set focus on input
    if (current) {
      current.addEventListener("click", handleComponentClick);
    }

    // Clean up event listener to prevent memory leaks
    return () => {
      if (current) {
        current.removeEventListener("click", handleComponentClick);
      }
    };
  }, []);

  return (
    <div className={styles.searchAutosuggest} ref={container}>
      <div className={styles.iconWrapper}>
        <SearchIcon />
      </div>
      <div>
        {categories.map(category => {
          return (
            <Category
              key={category.value}
              category={category}
              onRemove={onRemoveCategory}
            />
          );
        })}
      </div>
      <div className={styles.inputWrapper}>
        <input
          ref={input}
          onChange={handleInputChange}
          placeholder={placeholder}
          type="text"
          value={searchValue}
        />
      </div>
    </div>
  );
};

export default SearchAutosuggest;
