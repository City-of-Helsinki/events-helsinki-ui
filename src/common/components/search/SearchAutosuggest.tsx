import { IconSearch } from 'hds-react';
import React, { ChangeEvent } from 'react';

import {
  AUTOSUGGEST_KEYWORD_BLACK_LIST,
  AUTOSUGGEST_TYPES,
} from '../../../constants';
import {
  LinkedEventsSource,
  useKeywordListQuery,
} from '../../../generated/graphql';
import useDebounce from '../../../hooks/useDebounce';
import useKeyboardNavigation from '../../../hooks/useDropdownKeyboardNavigation';
import useLocale from '../../../hooks/useLocale';
import getLocalisedString from '../../../util/getLocalisedString';
import { AutosuggestMenuOption } from '../../types';
import AutosuggestMenu from './AutosuggestMenu';
import styles from './searchAutosuggest.module.scss';

export interface SearchAutosuggestProps {
  name: string;
  onChangeSearchValue: (value: string) => void;
  onOptionClick: (item: AutosuggestMenuOption) => void;
  placeholder: string;
  searchValue: string;
  source?: LinkedEventsSource;
}

const SearchAutosuggest: React.FC<SearchAutosuggestProps> = ({
  name,
  onOptionClick,
  onChangeSearchValue,
  placeholder,
  searchValue,
  source,
}) => {
  const locale = useLocale();
  const container = React.useRef<HTMLDivElement | null>(null);
  const input = React.useRef<HTMLInputElement | null>(null);
  const [isMenuOpen, setIsMenuOpen] = React.useState(false);
  const internalInputValue = useDebounce(searchValue, 300);

  const { data: keywordsData, loading: loadingKeywords } = useKeywordListQuery({
    skip: !internalInputValue,
    variables: {
      hasUpcomingEvents: true,
      pageSize: 5,
      text: internalInputValue,
      source,
    },
  });

  const [autoSuggestItems, setAutoSuggestItems] = React.useState<
    AutosuggestMenuOption[]
  >([]);

  const ensureMenuIsClosed = React.useCallback(() => {
    if (isMenuOpen) {
      setIsMenuOpen(false);
    }
  }, [isMenuOpen]);

  const ensureMenuIsOpen = React.useCallback(() => {
    if (!isMenuOpen) {
      setIsMenuOpen(true);
    }
  }, [isMenuOpen]);

  const {
    focusedIndex,
    setup: setupKeyboardNav,
    teardown: teardownKeyboardNav,
  } = useKeyboardNavigation({
    container: container,
    listLength: autoSuggestItems.length,
    onKeyDown: (event: KeyboardEvent) => {
      switch (event.key) {
        case 'ArrowUp':
        case 'ArrowDown':
          ensureMenuIsOpen();
          break;
        case 'Escape':
          handleCloseMenu();
          break;
        case 'Enter':
          const selectedItem = autoSuggestItems[focusedIndex];

          if (selectedItem) {
            handleMenuOptionClick(selectedItem);
          } else {
            // Search by text if no option is selected
            handleMenuOptionClick({
              text: searchValue,
              type: AUTOSUGGEST_TYPES.TEXT,
              value: searchValue,
            });
          }
          break;
        case 'Tab':
          ensureMenuIsClosed();
      }
    },
  });

  React.useEffect(() => {
    if (loadingKeywords) return;

    const items: AutosuggestMenuOption[] = [];
    const textItem = {
      text: internalInputValue,
      type: AUTOSUGGEST_TYPES.TEXT,
      value: internalInputValue,
    };

    items.push(textItem);

    items.push(
      ...(keywordsData?.keywordList.data
        .filter((keyword) => {
          const name = getLocalisedString(keyword.name, locale).toLowerCase();
          return (
            !AUTOSUGGEST_KEYWORD_BLACK_LIST.includes(keyword.id || '') &&
            name &&
            name !== textItem.text.toLowerCase()
          );
        })
        .map((keyword) => ({
          text: getLocalisedString(keyword.name, locale),
          type: AUTOSUGGEST_TYPES.KEYWORD,
          value: keyword.id || '',
        })) || [])
    );

    setAutoSuggestItems(items.filter((item) => item.text));
  }, [internalInputValue, keywordsData, loadingKeywords, locale]);

  const handleCloseMenu = React.useCallback(() => {
    setFocusToInput();

    ensureMenuIsClosed();
  }, [ensureMenuIsClosed]);

  const handleMenuOptionClick = React.useCallback(
    (option: AutosuggestMenuOption) => {
      onOptionClick(option);
      handleCloseMenu();
    },
    [handleCloseMenu, onOptionClick]
  );

  const setFocusToInput = () => {
    input.current?.focus();
  };

  const handleComponentClick = () => {
    setFocusToInput();
  };

  const onDocumentClick = (event: MouseEvent) => {
    const target = event.target;
    const current = container.current;

    // Close menu when clicking outside of the component
    if (!(target instanceof Node && current?.contains(target))) {
      setIsMenuOpen(false);
    }
  };

  const onDocumentFocusin = React.useCallback(
    (event: FocusEvent) => {
      const target = event.target;
      const current = container.current;

      if (!(target instanceof Node && current?.contains(target))) {
        ensureMenuIsClosed();
      }
    },
    [ensureMenuIsClosed]
  );

  const handleInputChange = (event: ChangeEvent<HTMLInputElement>) => {
    const newValue = event.target.value;

    onChangeSearchValue(newValue);

    ensureMenuIsOpen();
  };

  const handleInputFocus = () => {
    ensureMenuIsOpen();
  };

  React.useEffect(() => {
    setupKeyboardNav();
    document.addEventListener('click', onDocumentClick);
    document.addEventListener('focusin', onDocumentFocusin);
    // Clean up event listener to prevent memory leaks
    return () => {
      teardownKeyboardNav();
      document.removeEventListener('click', onDocumentClick);
      document.removeEventListener('focusin', onDocumentFocusin);
    };
  }, [onDocumentFocusin, setupKeyboardNav, teardownKeyboardNav]);

  return (
    <div
      onClick={handleComponentClick}
      className={styles.searchAutosuggest}
      ref={container}
    >
      <div className={styles.iconWrapper}>
        <IconSearch size="s" aria-hidden />
      </div>
      <div className={styles.inputWrapper}>
        <input
          ref={input}
          id={name}
          name={name}
          onChange={handleInputChange}
          onFocus={handleInputFocus}
          placeholder={placeholder}
          type="text"
          value={searchValue}
        />
      </div>
      <AutosuggestMenu
        focusedOption={focusedIndex}
        isOpen={!!searchValue && isMenuOpen}
        onClose={handleCloseMenu}
        onOptionClick={handleMenuOptionClick}
        options={autoSuggestItems}
      />
    </div>
  );
};

export default SearchAutosuggest;
