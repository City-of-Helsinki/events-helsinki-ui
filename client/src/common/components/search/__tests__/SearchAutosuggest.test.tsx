import * as React from "react";
import renderer from "react-test-renderer";

import SearchAutosuggest from "../SearchAutosuggest";

test("SearchAutosuggest matches snapshot", () => {
  const component = renderer.create(
    <SearchAutosuggest
      categories={[{ text: "bar", value: "foo" }]}
      onRemoveCategory={category => {
        // eslint-disable-next-line no-console
        console.log(category);
      }}
      placeholder="Lorem ipsum"
    />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

export {};
