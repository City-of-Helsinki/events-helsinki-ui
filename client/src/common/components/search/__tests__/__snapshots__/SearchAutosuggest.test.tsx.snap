// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchAutosuggest matches snapshot 1`] = `
<div
  className="searchAutosuggest"
>
  <div
    className="iconWrapper"
  >
    <svg>
      search.svg
    </svg>
  </div>
  <div>
    <div
      className="category"
    >
      <button
        aria-label="Poista kategoria: bar"
        className="closeButton"
        onClick={[Function]}
        type="button"
      >
        <svg>
          close.svg
        </svg>
      </button>
      bar
    </div>
  </div>
  <div
    className="inputWrapper"
  >
    <input
      onChange={[Function]}
      placeholder="Lorem ipsum"
      type="text"
      value=""
    />
  </div>
  <div
    className="autosuggestMenu"
  >
    <div
      className="title"
    >
      Hakuehdotuksia
    </div>
    <ul
      className="autosuggesItems"
    >
      <li
        className="autosuggesItem autosuggestItem--category"
      >
        <div
          className="colorIndicator"
        />
        <div
          className="icon"
        />
        <div
          className="textWrapper"
        >
          <button>
            <div
              className="text"
            >
              Luonto ja ulkoilu
            </div>
          </button>
        </div>
      </li>
      <li
        className="autosuggesItem autosuggestItem--yso"
      >
        <div
          className="colorIndicator"
        />
        <div
          className="icon"
        />
        <div
          className="textWrapper"
        >
          <button>
            <div
              className="text"
            >
              Maastojuoksu
            </div>
          </button>
        </div>
      </li>
      <li
        className="autosuggesItem autosuggestItem--yso"
      >
        <div
          className="colorIndicator"
        />
        <div
          className="icon"
        />
        <div
          className="textWrapper"
        >
          <button>
            <div
              className="text"
            >
              Maastopyöräily
            </div>
          </button>
        </div>
      </li>
      <li
        className="autosuggesItem autosuggestItem--yso"
      >
        <div
          className="colorIndicator"
        />
        <div
          className="icon"
        />
        <div
          className="textWrapper"
        >
          <button>
            <div
              className="text"
            >
              Sunnistus
            </div>
          </button>
        </div>
      </li>
      <li
        className="autosuggesItem autosuggestItem--area"
      >
        <div
          className="colorIndicator"
        />
        <div
          className="icon"
        />
        <div
          className="textWrapper"
        >
          <button>
            <div
              className="text"
            >
              Keskusta
            </div>
          </button>
        </div>
      </li>
      <li
        className="autosuggesItem autosuggestItem--area"
      >
        <div
          className="colorIndicator"
        />
        <div
          className="icon"
        />
        <div
          className="textWrapper"
        >
          <button>
            <div
              className="text"
            >
              Käpylä
            </div>
          </button>
        </div>
      </li>
      <li
        className="autosuggesItem autosuggestItem--servicePoint"
      >
        <div
          className="colorIndicator"
        />
        <div
          className="icon"
        />
        <div
          className="textWrapper"
        >
          <button>
            <div
              className="text"
            >
              Kisahalli
            </div>
          </button>
        </div>
      </li>
      <li
        className="autosuggesItem autosuggestItem--servicePoint"
      >
        <div
          className="colorIndicator"
        />
        <div
          className="icon"
        />
        <div
          className="textWrapper"
        >
          <button>
            <div
              className="text"
            >
              Annatalo
            </div>
          </button>
        </div>
      </li>
      <li
        className="autosuggesItem autosuggestItem--servicePoint"
      >
        <div
          className="colorIndicator"
        />
        <div
          className="icon"
        />
        <div
          className="textWrapper"
        >
          <button>
            <div
              className="text"
            >
              Stoa
            </div>
          </button>
        </div>
      </li>
      <li
        className="autosuggesItem autosuggestItem--servicePoint"
      >
        <div
          className="colorIndicator"
        />
        <div
          className="icon"
        />
        <div
          className="textWrapper"
        >
          <button>
            <div
              className="text"
            >
              Caisa
            </div>
          </button>
        </div>
      </li>
    </ul>
    <div
      className="info"
    >
      Voit myös etsiä kirjoittamallasi hakusanalla
    </div>
  </div>
</div>
`;
